GetStream/stream-chat-android
# [3881] Expose message position handler

### 🎯 Goal

Closes #3881 

Expose a handler to determine message position in a group: bottom, top or middle.

### 🎨 UI Changes

| Without the patch | With the patch applied |
| --- | --- |
| ![photo_2022-07-14 13 27 51](https://user-images.githubusercontent.com/9600921/178962232-b68ca470-337b-4e16-a94f-d4e2a9b31581.jpeg) | ![photo_2022-07-14 13 27 53](https://user-images.githubusercontent.com/9600921/178962225-58eaf549-732d-48e0-a685-1ead82e83325.jpeg) |

### 🧪 Testing

1. Apply the patch below
2. Send a message with "separator" text
3. Ensure that the message breaks the group

<details>

<summary>Provide the patch summary here</summary>

```
Index: stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/ChatFragment.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/ChatFragment.kt b/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/ChatFragment.kt
--- a/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/ChatFragment.kt	(revision ea7c58b14e791fcbeed39b50a287859893cb417b)
+++ b/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/chat/ChatFragment.kt	(date 1657790043412)
@@ -26,6 +26,7 @@
 import androidx.fragment.app.viewModels
 import androidx.navigation.fragment.findNavController
 import androidx.navigation.fragment.navArgs
+import com.getstream.sdk.chat.adapter.MessageListItem
 import com.getstream.sdk.chat.utils.extensions.getCreatedAtOrThrow
 import com.getstream.sdk.chat.viewmodel.MessageInputViewModel
 import com.getstream.sdk.chat.viewmodel.messages.MessageListViewModel
@@ -230,6 +231,32 @@
                     shouldShowDateSeparator(calendar, previousMessage, message)
                 }
             }
+            setMessagePositionHandler { prevMessage: Message?, message: Message, nextMessage: Message?, isAfterDateSeparator: Boolean ->
+                val prevUser = prevMessage?.user
+                val user = message.user
+                val nextUser = nextMessage?.user
+
+                fun Message.isServerMessage(): Boolean {
+                    return type == "system" || type == "error"
+                }
+
+                // Treat regular messages with certain extra data like system messages
+                fun Message.isSeparator(): Boolean {
+                    return extraData.containsKey("separator")
+                }
+
+                mutableListOf<MessageListItem.Position>().apply {
+                    if (prevMessage == null || prevUser != user || prevMessage.isServerMessage() || isAfterDateSeparator || prevMessage.isSeparator()) {
+                        add(MessageListItem.Position.TOP)
+                    }
+                    if (prevMessage != null && nextMessage != null && prevUser == user && nextUser == user) {
+                        add(MessageListItem.Position.MIDDLE)
+                    }
+                    if (nextMessage == null || nextUser != user || nextMessage.isServerMessage() || nextMessage.isSeparator()) {
+                        add(MessageListItem.Position.BOTTOM)
+                    }
+                }
+            }
             state.observe(viewLifecycleOwner) {
                 when (it) {
                     is MessageListViewModel.State.Loading -> Unit
Index: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt b/stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt
--- a/stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt	(revision ea7c58b14e791fcbeed39b50a287859893cb417b)
+++ b/stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt	(date 1657789838656)
@@ -1346,6 +1346,10 @@
                     interceptor.interceptMessage(channelType, channelId, message.data(), isRetrying)
                 } else message
             }.flatMapSuspend { newMessage ->
+                if (newMessage.text == "separator") {
+                    newMessage.extraData["separator"] = true
+                }
+
                 api.sendMessage(channelType, channelId, newMessage)
                     .retry(scope, retryPolicy)
                     .doOnResult(scope) { result ->

```

</details>


### ☑️Contributor Checklist

#### General

#### Code & documentation

### ☑️Reviewer Checklist
